lib/Parse/Lexer.cpp

```diff
@@ -1477,6 +1471,22 @@ void Lexer::lexImpl() {
       } else {
         diagnose(CurPtr-1, diag::lex_invalid_character)
           .fixItReplaceChars(getSourceLoc(CurPtr-1), getSourceLoc(tmp), " ");
+
+        llvm::SmallString<64> buffer;
+        EncodeToUTF8(codepoint, buffer);
+        StringRef confusedChar = buffer.str();
+        uint32_t expectedCodepoint;
+        if ((expectedCodepoint =
+            confusable::tryConvertConfusableCharacterToASCII(codepoint))) {
+          std::string expected = std::string(1, (char)expectedCodepoint);
+          StringRef expectedChar = StringRef(expected);
+          diagnose(CurPtr-1, diag::lex_confusable_character,
+                   confusedChar, expectedChar)
+            .fixItReplaceChars(getSourceLoc(CurPtr-1),
+                               getSourceLoc(tmp),
+                               expectedChar);
+        }
```

Comment:
Recovering here would mean jumping to the top of the switch with the expected character instead, but I'm not sure we're set up to handle that very well. We probably assume all over the place that all the ASCII characters are only one byte wide. I guess it's okay not to do anything here.